### ISSUE TITLE
Single file?

### ISSUE THREAD

**Issue by lordspace** [vizhub-core/vzcode#471]
is there a way to package this IDE as a single file so it can be uploaded in a site without the need of other dependencies?

**Comment by curran** [comment]
Great idea! It should be possible. The only real issue is that the way it's set up now, it needs a server to run the ShareDB part which synchronizes things to the file system.

Please take a look at what is generated by the `npm run build` command, and see if the generated files suit your needs.

Are you thinking of creating an editor that doesn't save the edited files anywhere? Or were you thinking of saving the files?

**Comment by lordspace** [comment]
Hi,
Thanks for the quick reply!
ideally, it would be great if I can use it as a file manager with some (basic) or user/pass/hash authentication and then interact with a PHP backend to edit files from any folders that the server has access to.

P.S. the UI in the project looks really nice. Forgot to mention that earlier.

**Comment by curran** [comment]
Interesting idea... This project has a pretty hard dependency on ShareDB, which is a Node-based server component that supports real-time collaboration.

I bet with some work, it would be possible to remove that ShareDB dependency for the front end, and instead have it fetch the initial files from an API on load, and call an API to save new versions of the files. That way you could use the PHP backend, but you'd lose the real-time collaboration feature (so if multiple people are editing, they can get out of sync and overwrite each other's work).

Perhaps the Node server could run as a peer to the PHP server, and requests could be proxied if needed? That way you'd keep the real-time features.

**Comment by curran** [comment]
Ahh I see! In that case, the ShareDB Node server would still be required, which is what synchronizes the changes between clients in real time and also writes changes to the file system.

That said, if having a Node server is not a possibility, I think it would be _in theory_ possible to refactor VZCode to interact with a REST API instead of a ShareDB server. We would just strip out all the ShareDB logic from the client, possibly by passing a feature flag boolean into the top level of the app or context.

If this is done, though, we would lose the real-time synchronization, so if multiple people are editing, or one person has the same thing open in two tabs, work could be lost. We'd probably want to introduce some sort of locking mechanism like Wikipedia does, where if the saved version on the server is newer than the one used to initialize your editor, you get a warning message and it won't let you save.

I'd be open to a contribution that does this refactoring.

To summarize the options:

 * You can use VZCode as-is, but that requires you run `npm run build` and upload the static files generated by the build (which is multiple files, not one), AND it requires that you have the ShareDB-based server running on your server to synchronize changes between clients and write files to disk
 * Additional work would be required to refactor the app to use a REST API for reading and writing files. This is possible, but would require implementation of that REST API, and would require a refactoring of the code to replace ShareDB-based saving with REST-based saving.

**Comment by lordspace** [comment]
Thanks @curran . I will do some planning in the next month or two and decide how to better approach this.
can I maybe copy some of tab CSS if it comes to that ? :smile:

**Comment by curran** [comment]
Good luck! I'd welcome a PR that untangles the ShareDB piece out of VZCode. Right now it's pretty tightly coupled.
